blog/_posts/2017-02-05-offset-arrays.md

```diff
@@ -0,0 +1,489 @@
+5×6 Array{Int64,2}:
+ 1   6  11  16  21  26
+ 2   7  12  17  22  27
+ 3   8  13  18  23  28
+ 4   9  14  19  24  29
+ 5  10  15  20  25  30
+```
+
+then we can refer to a rectangular region like this:
+
+```julia
+julia> B = A[1:3, 1:4]
+3×4 Array{Int64,2}:
+ 1  6  11  16
+ 2  7  12  17
+ 3  8  13  18
+```
+
+We can represent the same concept graphically using images:
+
+```julia
+using TestImages, Images
+
+img = testimage("mandrill")
+imgeye = img[35:99,130:220]
+imgnose = img[331:445,145:350]
+```
+
+If we display these three images, we get the following:
+
+![mandrill](https://github.com/JuliaLang/julialang.github.com/blob/master/blog/_posts/offsetarrays_figures/mandrill.png?raw=true)
+![mandrill](https://github.com/JuliaLang/julialang.github.com/blob/master/blog/_posts/offsetarrays_figures/mandrill_eye.png?raw=true)
+![mandrill](https://github.com/JuliaLang/julialang.github.com/blob/master/blog/_posts/offsetarrays_figures/mandrill_nose.png?raw=true)
+
+In summary, it's very straightforward to pick a particular region of
+an array and extract it for further analysis.  Like the Talking Heads
+song "Road to Nowhere," we know where we're going...
+
+## ...but we don't know where we've been...
+
+For certain applications, one negative to extracting blocks is that
+there is no record indicating where the new block originated from:
+
+```julia
+julia> B2 = A[2:4, 1:4]
+3×4 Array{Int64,2}:
+ 2  7  12  17
+ 3  8  13  18
+ 4  9  14  19
+
+julia> B2[1,1]
+2
+```
+
+So `B2[1,1]` corresponds to `A[2,1]`, despite the fact that, as
+measured by their indices, these are not the same location. Likewise,
+when snipping out regions of an image, we don't know where, relative
+to the original images, those pieces came from.  Often we may not
+care, but in other cases it's essential to pass forward both the
+snipped-out data *and* the indices used for the snipping. In complex
+cases--when you have indices of indices of indices--keeping track of
+location can make your head hurt.
+
+## ...give us time to work it out
+
+Starting with Julia 0.5, there is experimental support for a simple
+solution to such problems. In programming, one of the best ways to
+keep track of correspondences is to make sure that everything
+maintains consistent "naming," which in this case means having
+consistent indices.  We can do this using the [OffsetArrays](https://github.com/alsam/OffsetArrays.jl)
+package:
+
+```julia
+julia> using OffsetArrays
+
+julia> B3 = OffsetArray(A[2:4, 1:4], 2:4, 1:4)  # wrap the snipped-out piece in an OffsetArray
+OffsetArrays.OffsetArray{Int64,2,Array{Int64,2}} with indices 2:4×1:4:
+ 2  7  12  17
+ 3  8  13  18
+ 4  9  14  19
+
+julia> B3[3,4]
+18
+
+julia> A[3,4]
+18
+```
+
+So the indices in `B3` match those of `A`. Indeed, `B3` doesn't even
+have an element "named" `(1,1)`:
+
+```julia
+julia> B3[1,1]
+ERROR: BoundsError: attempt to access OffsetArrays.OffsetArray{Int64,2,Array{Int64,2}} with indices 2:4×1:4 at index [1,1]
```

Comment:
Good idea. Esp. since 0.5.0 (rather than release-0.5) has some significant bugs when it comes to `indices` support.